dcos_launch/terraform.py

```diff
@@ -0,0 +1,100 @@
+from dcos_launch import util
+import yaml
+import os
+import json
+import shutil
+import uuid
+
+# TODO pinning terraform-dcos is not enough because its tf_dcos_core dependency is not pinned. Create a branch on
+# terraform-dcos that pins tf_dcos_core and replace TERRAFORM_DCOS_VERSION here by that branch name. Regularly update
+# that branch with master.
+TERRAFORM_DCOS_VERSION = 'master'
+TERRAFORM_DCOS_EE_VERSION = 'service_credentials_gcp2'
+
+
+class TerraformLauncher(util.AbstractLauncher):
+    def __init__(self, config: dict, env=None):
+        self.config = config
+        self.init_dir = config.get('init_dir')
+        if not self.init_dir:
+            self.init_dir = 'terraform-init-' + uuid.uuid4().__str__()
+            self.config['init_dir'] = self.init_dir
+        self.init_dir = os.path.abspath(self.init_dir)
+        self.cluster_profile_path = os.path.join(self.init_dir, 'desired_cluster_profile.tfvars')
+        if env is None:
+            env = os.environ.copy()
+
+        if config['platform'] == 'azure':
+            # TODO
+            pass
+        elif config['platform'] == 'aws':
+            # TODO
+            pass
+        elif config['platform'] == 'gcp':
+            if 'GCE_CREDENTIALS' in env:
+                with open('gcp_service_creds.json', 'w') as file:
+                    file.write(env['GCE_CREDENTIALS'])
+                config['terraform_config']['gcp_credentials_key_file'] = os.path.abspath('gcp_service_creds.json')
+            elif 'GOOGLE_APPLICATION_CREDENTIALS' in env:
+                config['terraform_config']['gcp_credentials_key_file'] = env['GOOGLE_APPLICATION_CREDENTIALS']
+            else:
+                raise util.LauncherError(
+                    'MissingParameter', 'Either GCE_CREDENTIALS or GOOGLE_APPLICATION_CREDENTIALS must be set in env')
+
+            if 'gcp_project' not in config['terraform_config']:
+                config['terraform_config']['gcp_project'] = json.load(
+                    config['terraform_config']['gcp_credentials_key_file'])['project_id']
+        else:
+            raise util.LauncherError('InvalidPlatform', "allowed values: 'gcp', 'aws' and 'azure'")
+
+        # create terraform directory
+        if not os.path.exists(self.init_dir):
+            os.makedirs(self.init_dir)
+            module = 'github.com/dcos/{}?ref={}/{}'.format(
+                'terraform-dcos-enterprise' if config['dcos-enterprise'] else 'terraform-dcos',
+                TERRAFORM_DCOS_EE_VERSION if config['dcos-enterprise'] else TERRAFORM_DCOS_VERSION,
+                config['platform'])
+            util.run_subprocess('terraform init -from-module ' + module, self.init_dir)
+
+            with open(self.cluster_profile_path, 'w') as file:
```

Comment:
The formatting work in this clause is a little complicated. Could you add a link in a comment to where the spec for this is?

Reply:
Done